# **Importer les Bibliothèques et Initialiser Spark**

In [46]:
import sys
from pyspark.sql import SparkSession
import matplotlib.pyplot as plt
from pyspark.sql import functions as F
import seaborn as sns
from pyspark.sql.types import StructType, StructField, StringType, FloatType, IntegerType, BooleanType

In [47]:
spark = SparkSession.builder \
    .appName("Data Processing dans catalogue csv") \
    .config("spark.hadoop.hive.metastore.uris", "thrift://hive-metastore:9083") \
    .enableHiveSupport() \
    .getOrCreate()
print("Session Spark initialisée avec succès.")


Session Spark initialisée avec succès.


# **Charger et normaliser  les données depuis le fichier Hive**


In [48]:
# Charger les données depuis Hive
spark.sql("USE concessionnaire")
catalogue_df = spark.sql("SELECT * FROM catalogue_table_processed")
co2_df = spark.sql("SELECT * FROM co2_data_processed")


# Normaliser les colonnes de marque
co2_df = co2_df.withColumn("marque", F.lower(F.trim(F.col("marque"))))
catalogue_df = catalogue_df.withColumn("marque", F.lower(F.trim(F.col("marque"))))

co2_df = co2_df.withColumn("modele", F.lower(F.trim(F.col("modele"))))
catalogue_df = catalogue_df.withColumn("modele", F.lower(F.trim(F.col("modele"))))
# Vérifier les données chargées
print("Données Catalogue :")
catalogue_df.show(5)

print("Données CO2 :")
co2_df.show(1000)

Données Catalogue :
+----------+----------+-------------+-------------------+-----------+--------+--------+-------+--------+--------+
|    marque|    modele|modele_detail|unified_horse_power|   longueur|nbplaces|nbportes|couleur|occasion|    prix|
+----------+----------+-------------+-------------------+-----------+--------+--------+-------+--------+--------+
|   renault|    espace|         2.0T|                165|     longue|       7|       5|  rouge|    true| 21245.0|
|volkswagen|new beatle|          1.8|                110|    moyenne|       5|       5|   gris|   false| 26630.0|
|       bmw|        m5|             |                507|très longue|       5|       5|   bleu|    true| 66360.0|
|    nissan|    maxima|       3.0 V6|                200|très longue|       5|       5|  rouge|   false| 30000.0|
|  mercedes|      s500|             |                306|très longue|       5|       5|   noir|   false|101300.0|
+----------+----------+-------------+-------------------+-----------

# **Creation de **final_catalogue_df**  qui servira lors de la fusion des data  **

In [49]:
# Définir le schéma du DataFrame vide avec les bonnes colonnes
schema = StructType([
    StructField("marque", StringType(), True),
    StructField("modele", StringType(), True),
    StructField("modele_detail", StringType(), True),
    StructField("unified_horse_power", FloatType(), True),
    StructField("longueur", FloatType(), True),
    StructField("nbplaces", IntegerType(), True),
    StructField("nbportes", IntegerType(), True),
    StructField("couleur", StringType(), True),
    StructField("occasion", BooleanType(), True),
    StructField("prix", FloatType(), True),
    StructField("bonus_malus", FloatType(), True),
    StructField("cout_energie", FloatType(), True),
    StructField("rejets_co2", FloatType(), True)
])

# Créer le DataFrame vide avec ce schéma
final_catalogue_df = spark.createDataFrame([], schema)

# Afficher le DataFrame vide
final_catalogue_df.show(truncate=False)


+------+------+-------------+-------------------+--------+--------+--------+-------+--------+----+-----------+------------+----------+
|marque|modele|modele_detail|unified_horse_power|longueur|nbplaces|nbportes|couleur|occasion|prix|bonus_malus|cout_energie|rejets_co2|
+------+------+-------------+-------------------+--------+--------+--------+-------+--------+----+-----------+------------+----------+
+------+------+-------------+-------------------+--------+--------+--------+-------+--------+----+-----------+------------+----------+


# **Drop duplicate sur co2 et catalogue**

In [50]:
# Liste des colonnes à utiliser pour vérifier l'unicité (toutes sauf 'couleur')
columns_to_check = [col for col in catalogue_df.columns if col != "couleur"]

# Supprimer doublons dans le fichier CO2 au cas ou
co2_df = co2_df.dropDuplicates()

# Supprimer les doublons en se basant sur toutes les colonnes sauf 'couleur'
catalogue_df = catalogue_df.dropDuplicates(columns_to_check)

# Vérifier les résultats après suppression des doublons
print(f"Nombre de lignes uniques dans Co2 après suppression des doublons : {co2_df.count()}")
print(f"Nombre de lignes uniques dans Catalogue après suppression des doublons (sans 'couleur') : {catalogue_df.count()}")
catalogue_df.show(truncate=False)


Nombre de lignes uniques dans Co2 après suppression des doublons : 175
Nombre de lignes uniques dans Catalogue après suppression des doublons (sans 'couleur') : 54
+----------+----------+-------------+-------------------+--------+--------+--------+-------+--------+-------+
|marque    |modele    |modele_detail|unified_horse_power|longueur|nbplaces|nbportes|couleur|occasion|prix   |
+----------+----------+-------------+-------------------+--------+--------+--------+-------+--------+-------+
|volkswagen|polo      |1.2 6V       |55                 |courte  |5       |3       |blanc  |true    |8540.0 |
|volkswagen|polo      |1.2 6V       |55                 |courte  |5       |3       |blanc  |false   |12200.0|
|daihatsu  |cuore     |1.0          |58                 |courte  |5       |3       |noir   |false   |8850.0 |
|kia       |picanto   |1.1          |65                 |courte  |5       |5       |blanc  |false   |8990.0 |
|audi      |a2        |1.4          |75                 |courte  |

# **Fusionner les Données dans final_catalogue_df**

In [51]:
#fusionner les données
final_catalogue_df = final_catalogue_df.unionByName(catalogue_df, allowMissingColumns=True)

print(f"Nombre de lignes dans le catalogue final fusionné : {final_catalogue_df.count()}")
final_catalogue_df.show()

Nombre de lignes dans le catalogue final fusionné : 54
+----------+----------+-------------+-------------------+--------+--------+--------+-------+--------+-------+-----------+------------+----------+
|    marque|    modele|modele_detail|unified_horse_power|longueur|nbplaces|nbportes|couleur|occasion|   prix|bonus_malus|cout_energie|rejets_co2|
+----------+----------+-------------+-------------------+--------+--------+--------+-------+--------+-------+-----------+------------+----------+
|volkswagen|      polo|       1.2 6V|               55.0|  courte|       5|       3|  blanc|    true| 8540.0|       NULL|        NULL|      NULL|
|volkswagen|      polo|       1.2 6V|               55.0|  courte|       5|       3|  blanc|   false|12200.0|       NULL|        NULL|      NULL|
|  daihatsu|     cuore|          1.0|               58.0|  courte|       5|       3|   noir|   false| 8850.0|       NULL|        NULL|      NULL|
|       kia|   picanto|          1.1|               65.0|  courte|   

# **Modèles du catalogue absent de co2 et ajout des m**

In [52]:
# Étape 1 : Identifiez les modèles absents
catalogue_missing_in_co2 = catalogue_df.join(
    co2_df.select("marque", "modele"),  # Sélectionne les colonnes nécessaires
    on=["marque", "modele"],
    how="left_anti"  # Garde uniquement les modèles absents dans CO2
)
catalogue_missing_in_co2.createOrReplaceTempView("missing_models")
print(f"Nombre de modèles absents dans CO2 : {catalogue_missing_in_co2.count()}")

# Étape 2 : Calculez les moyennes par marque dans CO2
averages_by_marque = co2_df.groupBy("marque").agg(
    F.round(F.avg("bonus_malus"), 1).alias("avg_bonus_malus"),
    F.round(F.avg("rejets_co2"), 1).alias("avg_rejets_co2"),
    F.round(F.avg("cout_energie"), 1).alias("avg_cout_energie")
)

averages_by_marque.createOrReplaceTempView("average_by_marque")

# Étape 3 : Associez les moyennes aux modèles absents
models_with_averages = spark.sql("""
    SELECT 
        m.marque, 
        m.modele,
        m.modele_detail,
        m.unified_horse_power,
        m.longueur,
        m.nbplaces,
        m.nbportes,
        m.couleur,
        m.occasion,
        m.prix,
        a.avg_bonus_malus AS bonus_malus,
        a.avg_rejets_co2 AS rejets_co2,
        a.avg_cout_energie AS cout_energie
    FROM missing_models m
    LEFT JOIN average_by_marque a
    ON m.marque = a.marque
""")


spark.catalog.dropTempView("average_by_marque")
print("Résultats des modèles absents enrichis avec les moyennes :")
models_with_averages.show(truncate=False)
print(f"Nombre de modèles absents dans CO2 enrichis avec les moyennes : {models_with_averages.count()}")

# Étape 4 : Fusionnez les données avec le catalogue initial
final_catalogue_df = final_catalogue_df.unionByName(
    models_with_averages,allowMissingColumns=True
)

# Étape 5 : Affichez les résultats finaux
print("Catalogue final enrichi avec les moyennes de CO2 pour les modèles absents :")
final_catalogue_df.show(1000)

#nombre de ligne dans final_catalogue_df
print(f"Nombre de lignes dans le catalogue final enrichi : {final_catalogue_df.count()}")





Nombre de modèles absents dans CO2 : 49
Résultats des modèles absents enrichis avec les moyennes :
+----------+----------+-------------+-------------------+--------+--------+--------+-------+--------+-------+-----------+----------+------------+
|marque    |modele    |modele_detail|unified_horse_power|longueur|nbplaces|nbportes|couleur|occasion|prix   |bonus_malus|rejets_co2|cout_energie|
+----------+----------+-------------+-------------------+--------+--------+--------+-------+--------+-------+-----------+----------+------------+
|volkswagen|polo      |1.2 6V       |55                 |courte  |5       |3       |blanc  |true    |8540.0 |-6000.0    |13.3      |149.3       |
|volkswagen|polo      |1.2 6V       |55                 |courte  |5       |3       |blanc  |false   |12200.0|-6000.0    |13.3      |149.3       |
|daihatsu  |cuore     |1.0          |58                 |courte  |5       |3       |noir   |false   |8850.0 |NULL       |NULL      |NULL        |
|kia       |picanto   |1.

# **Recalculcul des moyenne dasn final_df  et co2**


In [53]:
# Enregistrer final_catalogue_df comme vue temporaire
final_catalogue_df.createOrReplaceTempView("final_catalogue_temp_view")
co2_df.createOrReplaceTempView("co2_temp_view")


# Combiner les deux sources et calculer les moyennes
combined_averages_sql = spark.sql("""
    SELECT 
        ROUND(AVG(bonus_malus), 1) AS avg_bonus_malus,
        ROUND(AVG(rejets_co2), 1) AS avg_rejets_co2,
        ROUND(AVG(cout_energie), 1) AS avg_cout_energie
    FROM (
        SELECT bonus_malus, rejets_co2, cout_energie 
        FROM final_catalogue_temp_view
        WHERE bonus_malus IS NOT NULL AND rejets_co2 IS NOT NULL AND cout_energie IS NOT NULL
        UNION ALL
        SELECT bonus_malus, rejets_co2, cout_energie 
        FROM co2_temp_view
        WHERE bonus_malus IS NOT NULL AND rejets_co2 IS NOT NULL AND cout_energie IS NOT NULL
    ) combined_data
""")
combined_averages = combined_averages_sql.collect()[0]
print("Moyennes combinées (final_catalogue_df et co2_df) :", combined_averages)

print({
    "avg_bonus_malus": combined_averages["avg_bonus_malus"],
    "avg_rejets_co2": combined_averages["avg_rejets_co2"],
    "avg_cout_energie": combined_averages["avg_cout_energie"]
})


#drop temp view
spark.catalog.dropTempView("final_catalogue_temp_view")
spark.catalog.dropTempView("co2_temp_view")




Moyennes combinées (final_catalogue_df et co2_df) : Row(avg_bonus_malus=1182.3, avg_rejets_co2=103.7, avg_cout_energie=513.6)
{'avg_bonus_malus': 1182.3, 'avg_rejets_co2': 103.7, 'avg_cout_energie': 513.6}


True

# **Marques Absentes dans CO2**

In [54]:
# Étape 1 : Identifiez les marques absentes dans `co2_df`
marques_missing_in_co2 = catalogue_df.select("marque").distinct().join(
    co2_df.select("marque").distinct(), 
    on="marque",
    how="left_anti"
)
marques_missing_in_co2.createOrReplaceTempView("marques_missing")

print("Marques dans le Catalogue mais absentes dans CO2 :")
marques_missing_in_co2.show(truncate=False)
print(f"Nombre de marques dans Catalogue mais absentes dans CO2 : {marques_missing_in_co2.count()}")

# Étape 2 : Récupérez les lignes correspondantes de `catalogue_df`
catalogue_marques_absentes = spark.sql("""
    SELECT 
        c.marque,
        c.modele,
        c.modele_detail,
        c.unified_horse_power,
        c.longueur,
        c.nbplaces,
        c.nbportes,
        c.couleur,
        c.occasion,
        c.prix
    FROM catalogue_table_processed c
    WHERE c.marque IN (SELECT marque FROM marques_missing)
""")

# Supprimer la vue temporaire `marques_missing`
spark.catalog.dropTempView("marques_missing")

print("Lignes du catalogue pour les marques absentes dans CO2 :")
catalogue_marques_absentes.show(truncate=False)

# Étape 3 : Ajouter les moyennes combinées pour les marques absentes
catalogue_marques_absentes = catalogue_marques_absentes \
    .withColumn("bonus_malus", F.lit(combined_averages["avg_bonus_malus"])) \
    .withColumn("cout_energie", F.lit(combined_averages["avg_cout_energie"])) \
    .withColumn("rejets_co2", F.lit(combined_averages["avg_rejets_co2"])) \

# Vérifier les colonnes pour la fusion
print("Colonnes de final_catalogue_df :", final_catalogue_df.columns)
print("Colonnes de catalogue_marques_absentes :", catalogue_marques_absentes.columns)

# Étape 4 : Réorganiser les colonnes dans le même ordre
aligned_columns = [
    "marque", "modele", "modele_detail", "unified_horse_power",
    "longueur", "nbplaces", "nbportes", "couleur", "occasion",
    "prix", "bonus_malus",  "cout_energie","rejets_co2"
]

final_catalogue_df = final_catalogue_df.select(*aligned_columns)
catalogue_marques_absentes = catalogue_marques_absentes.select(*aligned_columns)

# Étape 5 : Fusionnez les données avec le `final_catalogue_df`
final_catalogue_df = final_catalogue_df.unionByName(catalogue_marques_absentes, allowMissingColumns=True)

# Résultats finaux
print("Catalogue final enrichi avec les moyennes pour les marques absentes :")
final_catalogue_df.show(truncate=False)
print(f"Nombre total de lignes dans le catalogue enrichi : {final_catalogue_df.count()}")

# Supprimer la vue temporaire `catalogue_table_processed` si nécessaire
spark.catalog.dropTempView("catalogue_table_processed")


Marques dans le Catalogue mais absentes dans CO2 :
+--------+
|marque  |
+--------+
|daihatsu|
|seat    |
|saab    |
|lancia  |
|hyundaï |
|ford    |
|honda   |
|fiat    |
|dacia   |
+--------+
Nombre de marques dans Catalogue mais absentes dans CO2 : 9
Lignes du catalogue pour les marques absentes dans CO2 :
+------+------+-------------+-------------------+--------+--------+--------+-------+--------+----+
|marque|modele|modele_detail|unified_horse_power|longueur|nbplaces|nbportes|couleur|occasion|prix|
+------+------+-------------+-------------------+--------+--------+--------+-------+--------+----+
+------+------+-------------+-------------------+--------+--------+--------+-------+--------+----+

Colonnes de final_catalogue_df : ['marque', 'modele', 'modele_detail', 'unified_horse_power', 'longueur', 'nbplaces', 'nbportes', 'couleur', 'occasion', 'prix', 'bonus_malus', 'cout_energie', 'rejets_co2']
Colonnes de catalogue_marques_absentes : ['marque', 'modele', 'modele_detail', 'unifie

ERROR:root:KeyboardInterrupt while sending command.
Traceback (most recent call last):
  File "/usr/local/spark/python/lib/py4j-0.10.9.7-src.zip/py4j/java_gateway.py", line 1038, in send_command
    response = connection.send_command(command)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/spark/python/lib/py4j-0.10.9.7-src.zip/py4j/clientserver.py", line 511, in send_command
    answer = smart_decode(self.stream.readline()[:-1])
                          ^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/socket.py", line 706, in readinto
    return self._sock.recv_into(b)
           ^^^^^^^^^^^^^^^^^^^^^^^
KeyboardInterrupt


KeyboardInterrupt: 

In [ ]:
spark.stop()

# **Vérifier les Données Fusionnées**

# **Enregistrer les Données Fusionnées**

In [ ]:
# # Enregistrer les données dans Hive
# catalogue_with_co2.write.mode("overwrite").saveAsTable("catalogue_co2_merge_processed")
# print("Données fusionnées enregistrées avec succès.")

# **Analyser les Données Fusionnées**

In [ ]:
# catalogue_with_co2.describe(["prix", "rejets_co2", "unified_horse_power", "longueur", "nbplaces"]).show()
# 
# # Vérification des véhicules avec rejets CO2 = 0
# zero_co2_vehicles = catalogue_with_co2.filter(F.col("rejets_co2") == 0)
# 
# # Compter le nombre total de véhicules avec rejets CO2 = 0
# zero_co2_count = zero_co2_vehicles.count()
# total_vehicles = catalogue_with_co2.count()
# 
# # Compter les véhicules avec rejets CO2 = 0 et un malus
# zero_co2_malus_count = zero_co2_vehicles.filter(F.col("bonus_malus") < 0).count()
# 
# # Calculer les proportions
# zero_co2_proportion = zero_co2_count / total_vehicles * 100
# zero_co2_malus_proportion = zero_co2_malus_count / zero_co2_count * 100 if zero_co2_count > 0 else 0
# 
# # Afficher les résultats
# print(f"Nombre total de véhicules : {total_vehicles}")
# print(f"Nombre de véhicules avec rejets CO2 = 0 : {zero_co2_count}")
# print(f"Proportion de véhicules avec rejets CO2 = 0 : {zero_co2_proportion:.2f}%")
# print(f"Nombre de véhicules avec rejets CO2 = 0 et malus : {zero_co2_malus_count}")
# print(f"Proportion des véhicules avec rejets CO2 = 0 et malus : {zero_co2_malus_proportion:.2f}%")
# 
# # Afficher les détails des incohérences
# print("Détails des véhicules avec rejets CO2 = 0 et malus :")
# zero_co2_vehicles.filter(F.col("bonus_malus") < 0).select(
#     "marque", "modele", "rejets_co2", "bonus_malus", "prix"
# ).show()
# 
# 
# 
# # Identifier les véhicules avec rejets CO2 = 0
# zero_co2_vehicles = catalogue_with_co2.filter(F.col("rejets_co2") == 0)
# 
# # Vérifier les marques concernées
# zero_co2_vehicles.groupBy("marque").count().show()
# 
# # Détail des modèles avec rejets CO2 = 0 et malus
# zero_co2_malus_vehicles = zero_co2_vehicles.filter(F.col("bonus_malus") < 0)
# zero_co2_malus_vehicles.select("marque", "modele", "rejets_co2", "bonus_malus").show(100)
# 


# **Données manquantes**

In [ ]:
# from pyspark.sql.functions import col, sum, when
# null_counts = co2_df.select([sum(when(col(c).isNull(), 1).otherwise(0)).alias(c) for c in co2_df.columns])
# null_counts.show()


# **Distribution des Prix des Véhicules**

In [ ]:
# prices = catalogue_with_co2.select("prix").rdd.flatMap(lambda x: x).collect()
# plt.hist(prices, bins=20, color='blue', edgecolor='black')
# plt.xlabel("Prix (€)")
# plt.ylabel("Fréquence")
# plt.title("Distribution des Prix des Véhicules")
# plt.grid(axis='y', linestyle='--')
# plt.show()


La majorité des véhicules ont un prix situé entre 20 000 € et 40 000 €, avec un pic autour de 25 000 €. Les véhicules de plus de 60 000 € sont peu nombreux, ce qui montre une distribution asymétrique centrée sur des véhicules de milieu de gamme.

# **Distribution de la Puissance des Véhicules**

In [ ]:
# horsepower_data = catalogue_with_co2.select("unified_horse_power").rdd.flatMap(lambda x: x).collect()
# plt.hist(horsepower_data, bins=20, color='green', edgecolor='black')
# plt.xlabel("Puissance (ch)")
# plt.ylabel("Fréquence")
# plt.title("Distribution de la Puissance des Véhicules")
# plt.grid(axis='y', linestyle='--')
# plt.show()


La majorité des véhicules ont une puissance comprise entre 100 ch et 200 ch, avec un pic autour de 150 ch. Les véhicules de plus de 300 ch sont rares, ce qui indique une concentration sur des voitures de puissance moyenne.

# **Répartition des Marques**

In [ ]:
# brand_distribution = catalogue_with_co2.groupBy("marque").count().toPandas()
# brand_distribution["percentage"] = (brand_distribution["count"] / brand_distribution["count"].sum()) * 100
# 
# plt.figure(figsize=(8, 8))
# plt.pie(brand_distribution["count"], startangle=90, colors=plt.cm.Paired.colors)
# labels = [f"{row['marque']} ({row['percentage']:.1f}%)" for _, row in brand_distribution.iterrows()]
# plt.legend(labels, title="Marques", loc="center left", bbox_to_anchor=(1, 0.5), fontsize=10)
# plt.title("Répartition des Marques")
# plt.tight_layout()
# plt.show()


# **Matrice de Corrélation**

In [ ]:
# correlations = catalogue_with_co2.select("prix", "bonus_malus", "rejets_co2", "unified_horse_power").toPandas().corr()
# sns.heatmap(correlations, annot=True, cmap="coolwarm")
# plt.title("Matrice de Corrélation")
# plt.show()


Le prix est fortement corrélé à la puissance unifiée des véhicules (0.87), ce qui est attendu pour des voitures plus puissantes.
Les rejets de CO₂ sont très fortement corrélés au bonus/malus (0.94), reflétant leur impact direct sur la taxe écologique.
La puissance unifiée a une très faible corrélation négative avec les rejets de CO₂ (-0.09), probablement dû à la présence de véhicules électriques ou hybrides puissants mais peu polluants.

# **Relation entre Puissance et Prix**

In [ ]:
# price_hp_data = catalogue_with_co2.select("prix", "unified_horse_power").rdd.map(tuple).collect()
# prices, hp = zip(*price_hp_data)
# plt.scatter(hp, prices, alpha=0.7, color='purple', edgecolor='black')
# plt.xlabel("Puissance (ch)")
# plt.ylabel("Prix (€)")
# plt.title("Relation entre Puissance et Prix des Véhicules")
# plt.grid(True)
# plt.show()


Le prix des véhicules augmente généralement avec la puissance, avec une tendance claire pour les puissances jusqu'à 200 ch. Cependant, au-delà de 300 ch, les prix varient fortement, reflétant la diversité des segments de véhicules, notamment les modèles haut de gamme

# **Relation entre Rejets CO2 et Bonus/Malus**

In [ ]:
# df_pandas = catalogue_with_co2.select("rejets_co2", "bonus_malus").toPandas()
# sns.regplot(x="rejets_co2", y="bonus_malus", data=df_pandas, scatter_kws={"alpha": 0.6}, line_kws={"color": "red"})
# plt.xlabel("Rejets CO2 (g/km)")
# plt.ylabel("Bonus/Malus (€)")
# plt.title("Relation entre Rejets CO2 et Bonus/Malus")
# plt.grid(True)
# plt.show()


Une relation linéaire claire existe entre les rejets de CO₂ et le bonus/malus. Les véhicules avec des émissions élevées subissent un malus (valeurs négatives), tandis que ceux avec des émissions très faibles bénéficient d'un bonus (valeurs positives).

# **Analyse des Bonus/Malus**

In [ ]:
# co2_df = catalogue_with_co2.withColumn("bonus_malus_category", 
#     F.when(F.col("bonus_malus") > 0, "Bonus")
#     .when(F.col("bonus_malus") < 0, "Malus")
#     .otherwise("Neutre ou Inconnu"))
# 
# co2_df.groupBy("bonus_malus_category").count().show()
# 
# # Visualisation
# bonus_dist = co2_df.groupBy("bonus_malus_category").count().toPandas()
# plt.pie(bonus_dist["count"], labels=bonus_dist["bonus_malus_category"], autopct="%1.1f%%")
# plt.title("Répartition des Catégories Bonus/Malus")
# plt.show()


# **Analyse des Marques**

In [ ]:
# catalogue_with_co2.groupBy("marque").agg(F.avg("unified_horse_power").alias("avg_hp")).orderBy(F.desc("avg_hp")).show(10)


# **Valeurs aberrantes**

In [ ]:
# from pyspark.sql.functions import mean, stddev
# 
# stats = catalogue_with_co2.select(mean("unified_horse_power").alias("mean"), stddev("unified_horse_power").alias("stddev")).first()
# mean_hp, stddev_hp = stats["mean"], stats["stddev"]
# 
# outliers = catalogue_with_co2.filter(
#     (F.col("unified_horse_power") > mean_hp + 3 * stddev_hp) |
#     (F.col("unified_horse_power") < mean_hp - 3 * stddev_hp)
# )
# print(f"Nombre de valeurs aberrantes : {outliers.count()}")
# outliers.select("marque", "modele", "unified_horse_power", "prix").show()


Parmi les valeurs aberrantes, on observe des véhicules comme la BMW M5, qui combine une puissance très élevée (507 ch) avec des prix particulièrement élevés (>60 000 €). Ces valeurs reflètent des modèles de luxe ou de performance qui se situent bien au-delà de la moyenne des données.

In [ ]:
# # Ajouter une colonne de catégorie personnalisée
# co2_df = co2_df.withColumn(
#     "categorie_vehicule",
#     F.when(
#         (F.col("prix") < 20000) & 
#         (F.col("unified_horse_power") < 100) & 
#         (F.col("rejets_co2") < 50) & 
#         (F.col("bonus_malus") >= 0),
#         "Petite voiture économique"
#     ).when(
#         (F.col("longueur") > 4.5) & 
#         (F.col("nbplaces") >= 5) & 
#         (F.col("prix").between(30000, 50000)) & 
#         (F.col("unified_horse_power").between(150, 250)) & 
#         (F.col("rejets_co2").between(50, 150)),
#         "SUV familial"
#     ).when(
#         (F.col("prix") > 50000) & 
#         (F.col("unified_horse_power") > 300) & 
#         (F.col("rejets_co2") > 150) & 
#         (F.col("longueur") > 4.8),
#         "Voiture de luxe ou de sport"
#     ).when(
#         (F.col("unified_horse_power") > 200) & 
#         (F.col("rejets_co2") < 50) & 
#         (F.col("bonus_malus") > 0) & 
#         (F.col("prix") > 40000),
#         "Hybride ou électrique haut de gamme"
#     ).when(
#         (F.col("prix").between(20000, 40000)) & 
#         (F.col("unified_horse_power").between(100, 200)) & 
#         (F.col("rejets_co2").between(50, 150)),
#         "Véhicule polyvalent standard"
#     ).when(
#         (F.col("prix") > 80000) & 
#         (F.col("unified_horse_power") > 400) & 
#         (F.col("rejets_co2") > 150),
#         "Véhicule ultra-sportif"
#     ).when(
#         (F.col("prix") < 30000) & 
#         (F.col("unified_horse_power") < 150) & 
#         (F.col("rejets_co2") < 50) & 
#         (F.col("bonus_malus") > 0),
#         "Véhicule éco-responsable entrée de gamme"
#     ).otherwise("Non catégorisé")  # Catégorie par défaut si aucun critère n'est rempli
# )
# 
# # Afficher quelques exemples pour vérifier les catégories
# co2_df.select("marque", "modele", "prix", "unified_horse_power", "rejets_co2", "bonus_malus", "categorie_vehicule").show(200)


# **Fermer la session Spark**


In [ ]:
# Arrêter la session Spark
# spark.stop()